File: experiments/repair-loop/testkits/003-rate-limiter/python/solution.py
"""Phase-1.75 stub token-bucket rate-limiter solution for Python adapter validation.

Minimal working implementation: satisfies the 003-rate-limiter API surface
with correct types and basic invariants, passes pyright (strict-mode off)
and the colocated harness tests. Not the canonical Phase-2/3 solution —
agents will write their own.
"""
from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class LimiterState:
    capacity: int
    tokens: int
    refill_rate: int
    last_tick: int


class LimiterError(ValueError):
    """Base class for limiter construction failures."""


class NonPositiveCapacityError(LimiterError):
    pass


class NegativeRefillRateError(LimiterError):
    pass


def new_limiter(capacity: int, refill_rate: int) -> LimiterState:
    if capacity <= 0:
        raise NonPositiveCapacityError(capacity)
    if refill_rate < 0:
        raise NegativeRefillRateError(refill_rate)
    return LimiterState(
        capacity=capacity,
        tokens=capacity,
        refill_rate=refill_rate,
        last_tick=0,
    )


def allow(state: LimiterState, tick: int) -> tuple[LimiterState, bool]:
    delta_ticks = tick - state.last_tick
    if delta_ticks < 0:
        delta_ticks = 0
    refilled = state.tokens + delta_ticks * state.refill_rate
    if refilled > state.capacity:
        refilled = state.capacity
    new_tick = tick if tick >= state.last_tick else state.last_tick
    if refilled >= 1:
        return (
            LimiterState(
                capacity=state.capacity,
                tokens=refilled - 1,
                refill_rate=state.refill_rate,
                last_tick=new_tick,
            ),
            True,
        )
    return (
        LimiterState(
            capacity=state.capacity,
            tokens=refilled,
            refill_rate=state.refill_rate,
            last_tick=new_tick,
        ),
        False,
    )


def tokens(state: LimiterState) -> int:
    return state.tokens

File: experiments/repair-loop/testkits/003-rate-limiter/go/solution.go
// Phase-1.75 stub token-bucket rate-limiter solution for Go adapter validation.
//
// Module-mode upgrade: paired with go.mod and solution_test.go (both
// harness-owned via targets/go.json `harness_files`). Verifier chain
// runs `go vet ./...`, `go build ./...`, `go test ./...`. Not the
// canonical Phase-2/3 solution — agents will write their own.
package main

import (
	"errors"
	"fmt"
)

type LimiterState struct {
	Capacity   int64
	Tokens     int64
	RefillRate int64
	LastTick   int64
}

var (
	ErrNonPositiveCapacity = errors.New("capacity must be positive")
	ErrNegativeRefillRate  = errors.New("refill_rate must be non-negative")
)

func NewLimiter(capacity, refillRate int64) (*LimiterState, error) {
	if capacity <= 0 {
		return nil, ErrNonPositiveCapacity
	}
	if refillRate < 0 {
		return nil, ErrNegativeRefillRate
	}
	return &LimiterState{
		Capacity:   capacity,
		Tokens:     capacity,
		RefillRate: refillRate,
		LastTick:   0,
	}, nil
}

func Allow(state *LimiterState, tick int64) (*LimiterState, bool) {
	deltaTicks := tick - state.LastTick
	if deltaTicks < 0 {
		deltaTicks = 0
	}
	refilled := state.Tokens + deltaTicks*state.RefillRate
	if refilled > state.Capacity {
		refilled = state.Capacity
	}
	newTick := tick
	if newTick < state.LastTick {
		newTick = state.LastTick
	}
	if refilled >= 1 {
		return &LimiterState{
			Capacity:   state.Capacity,
			Tokens:     refilled - 1,
			RefillRate: state.RefillRate,
			LastTick:   newTick,
		}, true
	}
	return &LimiterState{
		Capacity:   state.Capacity,
		Tokens:     refilled,
		RefillRate: state.RefillRate,
		LastTick:   newTick,
	}, false
}

func Tokens(state *LimiterState) int64 {
	return state.Tokens
}

func main() {
	// Smoke main: not exercised by the harness. Exists only so `go build`
	// in package-main mode produces an executable.
	l, _ := NewLimiter(2, 1)
	l, ok := Allow(l, 0)
	fmt.Println("first allow:", ok, "tokens:", Tokens(l))
}
